### PR TITLE
feat(math): impl remove zeros for sparse univariate poly

### PR DESCRIPTION
**Pull request by batzor**
# Description

- Change `RemoveHighDegreeZeros()` to `RemoveZeros()` in univariate spare polynomial
- Enforce clean construction of polynomials. After this change, all poly operations are expected to operate on _clean_ polynomials and output _clean_ polynomials.

### CODE REVIEW

File: tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h [hunk 1]
```diff
@@ -104,13 +104,13 @@ class UnivariateSparseCoefficients {
   constexpr explicit UnivariateSparseCoefficients(
       const std::vector<Term>& terms)
       : terms_(terms) {
+    RemoveZeros();
     CHECK_LE(Degree(), kMaxDegree);
-    RemoveHighDegreeZeros();
```
Comment (chokobole) [inline tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h:108]: Why did you switch the order?

Comment (batzor) [reply tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h:108]: It will handle cases when the highest degree terms are 0.

Comment (batzor) [reply tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h:108]: To rephrase, Degree() can become less that kMaxDegree after RemoveZeros()

File: tachyon/math/polynomials/univariate/univariate_sparse_polynomial_unittest.cc [hunk 2]
```diff
@@ -324,6 +323,7 @@ TEST_F(UnivariateSparsePolynomialTest, MulScalar) {
   }
 
   Poly actual = poly * scalar;
+  actual.coefficients().RemoveZeros();
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/univariate/univariate_sparse_polynomial_unittest.cc:326]: Isn't `RemoveZeros()` called when multiplication is done? Do we need this code?

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

**Review by dongchangYoo — APPROVED** [review]
LGTM

File: tachyon/math/polynomials/univariate/univariate_sparse_polynomial_unittest.cc (hunk 2 above)
Comment (batzor) [reply tachyon/math/polynomials/univariate/univariate_sparse_polynomial_unittest.cc]: This test was failing in some random cases and I found that it fails when scalar is 0 the expected is constructed with ill-formed terms and cause the test to fail. I changed to following:

`expected.coefficients().RemoveZeros();`

**Review by Insun35 — APPROVED** [review]
LGTM

File: tachyon/math/polynomials/univariate/univariate_dense_coefficients.h [hunk 3]
```diff
@@ -69,8 +71,11 @@ class UnivariateDenseCoefficients {
   }
 
   constexpr static UnivariateDenseCoefficients Random(size_t degree) {
-    return UnivariateDenseCoefficients(
-        base::CreateVector(degree + 1, []() { return F::Random(); }));
+    auto v = base::CreateVector(degree + 1, []() { return F::Random(); });
```
Comment (ashjeong) [inline tachyon/math/polynomials/univariate/univariate_dense_coefficients.h:74]: Could you explicitly name the type as `std::vector<F>`?
```suggestion
    std::vector<F> v = base::CreateVector(degree + 1, []() { return F::Random(); });

**Review by ashjeong — CHANGES_REQUESTED** [review]

**Review by ashjeong — APPROVED** [review]
LGTM

File: tachyon/math/polynomials/univariate/univariate_dense_coefficients.h [hunk 4]
```diff
@@ -69,8 +71,12 @@ class UnivariateDenseCoefficients {
...
 
   constexpr static UnivariateDenseCoefficients Random(size_t degree) {
-    return UnivariateDenseCoefficients(
-        base::CreateVector(degree + 1, []() { return F::Random(); }));
+    std::vector v =
+        base::CreateVector(degree + 1, []() { return F::Random(); });
+    if (v[degree].IsZero()) {
+      v[degree] = F::One();
+    }
+    return UnivariateDenseCoefficients(std::move(v));
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/univariate/univariate_dense_coefficients.h:79]: Please make this change as a separate commit.

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

File: tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h [hunk 5]
```diff
@@ -102,15 +102,16 @@ class UnivariateSparseCoefficients {
 
   constexpr UnivariateSparseCoefficients() = default;
   constexpr explicit UnivariateSparseCoefficients(
-      const std::vector<Term>& terms)
+      const std::vector<Term>& terms, bool cleanup = false)
       : terms_(terms) {
+    if (cleanup) RemoveZeros();
     CHECK_LE(Degree(), kMaxDegree);
-    RemoveHighDegreeZeros();
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h:108]: This should've been removed in the previous commit.

File: tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h [hunk 6]
```diff
@@ -102,15 +102,16 @@ class UnivariateSparseCoefficients {
...
   }
-  constexpr explicit UnivariateSparseCoefficients(std::vector<Term>&& terms)
+  constexpr explicit UnivariateSparseCoefficients(std::vector<Term>&& terms,
+                                                  bool cleanup = false)
       : terms_(std::move(terms)) {
+    if (cleanup) RemoveZeros();
     CHECK_LE(Degree(), kMaxDegree);
-    RemoveHighDegreeZeros();
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h:113]: This should've been removed in the previous commit.

**Comment by TomTaehoonKim** [comment]
@batzor For the second commit, I can agree for the sparse coefficients but not with the dense coefficients. What's the motivation behind reducing the usage of remove high degree zeros for dense coefficients?

File: tachyon/math/polynomials/univariate/univariate_sparse_polynomial_unittest.cc [hunk 7]
```diff
@@ -40,11 +40,7 @@ TEST_F(UnivariateSparsePolynomialTest, IsZero) {
   EXPECT_TRUE(Poly().IsZero());
   EXPECT_TRUE(Poly::Zero().IsZero());
 
-  Poly p = Poly(Coeffs({{0, GF7(0)}, {1, GF7(0)}}));
-  p.coefficients().RemoveZeros();
-  EXPECT_TRUE(p.IsZero());
```
[the remark this answers is not in the document]
Comment (batzor) [reply tachyon/math/polynomials/univariate/univariate_sparse_polynomial_unittest.cc]: With this PR, we are enforcing polynomial construction with only _cleaned up_ coefficients. So this test fails in line 43 because the constructor checks if the coefficients are clean.

File: tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h (hunk 5 above)
Comment (batzor) [reply tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h:108]: This is the first commit :)

File: tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h (hunk 6 above)
Comment (batzor) [reply tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h:113]: This is in the first commit as well

File: tachyon/math/polynomials/univariate/univariate_dense_coefficients.h (hunk 4 above)
Comment (batzor) [reply tachyon/math/polynomials/univariate/univariate_dense_coefficients.h:79]: Umm may I ask why? I think it is part of enforcing proper polynomial construction. It wouldn't be a bug without this refactoring.

**Comment by batzor** [comment]
@TomTaehoonKim Indeed, the `RemoveHighDegreeZeros()` is low cost operation so it's not that necessary but I just did the change for the sake of consistency and clarity.

File: tachyon/math/polynomials/univariate/univariate_sparse_polynomial_unittest.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (batzor) [reply tachyon/math/polynomials/univariate/univariate_sparse_polynomial_unittest.cc]: I reverted the removal and fixed the test.

File: tachyon/math/polynomials/univariate/univariate_polynomial_ops.h [hunk 8]
```diff
@@ -766,6 +766,9 @@ class UnivariatePolynomialOp<UnivariateSparseCoefficients<F, MaxDegree>> {
   }
 
   static UnivariatePolynomial<D> ToDense(const UnivariatePolynomial<S>& self) {
+    if (self.IsZero()) {
+      return UnivariatePolynomial<D>::Zero();
+    }
```
[the remark this answers is not in the document]
Comment (batzor) [reply tachyon/math/polynomials/univariate/univariate_polynomial_ops.h:771]: Done :)

File: tachyon/math/polynomials/univariate/univariate_dense_coefficients.h (hunk 4 above)
Comment (batzor) [reply tachyon/math/polynomials/univariate/univariate_dense_coefficients.h:79]: Done :)

File: tachyon/math/polynomials/univariate/lagrange_interpolation.h [hunk 9]
```diff
@@ -101,7 +101,7 @@ bool LagrangeInterpolate(
       coeffs_sums[0][j] += coeffs_sums[i][j];
     }
   }
-  *ret = Poly(Coeffs(std::move(coeffs_sums[0])));
+  *ret = Poly(Coeffs(std::move(coeffs_sums[0]), true));
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/univariate/lagrange_interpolation.h:104]: What about cases like
1. https://github.com/kroma-network/tachyon/blob/e182278d97c58c934c3d81f00aa710df9c1c5115/tachyon/math/polynomials/univariate/lagrange_interpolation.h#L34
and
2. https://github.com/kroma-network/tachyon/blob/e182278d97c58c934c3d81f00aa710df9c1c5115/tachyon/zk/plonk/vanishing/vanishing_prover_impl.h#L178
?
I'm still not really convinced with "enforcing clean construction of polynomials.

File: tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h [hunk 10]
```diff
@@ -289,16 +294,6 @@ class UnivariateSparseCoefficients {
     return &it->coefficient;
   }
 
-  void RemoveHighDegreeZeros() {  // Fix to RemoveZeros
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h:292]: Have you tried build and tests? I'm pretty sure by removing this, it will fail because we call this when doing polynomial operations. https://github.com/kroma-network/tachyon/blob/e182278d97c58c934c3d81f00aa710df9c1c5115/tachyon/math/polynomials/univariate/univariate_polynomial_ops.h#L835

Comment (batzor) [reply tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h:292]: Yes I did build, test and lint.

Umm that seems to be from old commit? That line is removed in this commit if you look above.

File: tachyon/math/polynomials/univariate/lagrange_interpolation.h (hunk 9 above)
Comment (batzor) [reply tachyon/math/polynomials/univariate/lagrange_interpolation.h:104]: As I understood, some operations rely on input polynomial to be _clean_. 
For example, https://github.com/kroma-network/tachyon/blob/e182278d97c58c934c3d81f00aa710df9c1c5115/tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h#L227-L234

I think enforcing it will be more bug-free in the future. As for the cases you mentioned, I could recheck everything and pass `true` in the needed parts. I thought of making the dense polynomial to cleanup by default but still in your second case, we will have to pass `true` since the polynomial can be sparse.

**Comment by batzor** [comment]
@chokobole I reordered the commits as you requested. Hope its more clear now;;

File: tachyon/math/polynomials/univariate/lagrange_interpolation.h (hunk 9 above)
Comment (batzor) [reply tachyon/math/polynomials/univariate/lagrange_interpolation.h:104]: For now I added `true` to the cases you mentioned. I couldn't find any other instance.

File: tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h [hunk 11]
```diff
@@ -283,6 +283,13 @@ class UnivariateSparseCoefficients {
     return ss.str();
   }
 
+  void RemoveZeros() {
+    terms_.erase(
+        std::remove_if(terms_.begin(), terms_.end(),
+                       [](const Term& x) { return x.coefficient.IsZero(); }),
+        terms_.end());
```
Comment (chokobole) [inline tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h:290]:
```suggestion
    base::EraseIf(terms_, [](const Term& x) { return x.coefficient.IsZero(); }));
```

Comment (batzor) [reply tachyon/math/polynomials/univariate/univariate_sparse_coefficients.h]: done!

**Review by chokobole — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM
